src/operations/execute_operation.ts

```diff
@@ -188,111 +164,113 @@ function executeWithServerSelection<TResult>(
-      return callback(
-        new MongoServerError({
-          message: MMAPv1_RETRY_WRITES_ERROR_MESSAGE,
-          errmsg: MMAPv1_RETRY_WRITES_ERROR_MESSAGE,
-          originalError
-        })
-      );
+  if (session == null || !operation.hasAspect(Aspect.RETRYABLE)) {
+    // No session or non-retryable operation, simple early exit
+    try {
+      return await operation.executeAsync(server, session);
+    } finally {
+      if (session?.owner != null && session.owner === owner) {
+        await session.endSession().catch(() => null);
+      }
     }
+  }
 
-    if (isWriteOperation && !isRetryableWriteError(originalError)) {
-      return callback(originalError);
-    }
+  const willRetryRead = topology.s.options.retryReads && !inTransaction && operation.canRetryRead;
 
-    if (isReadOperation && !isRetryableReadError(originalError)) {
-      return callback(originalError);
-    }
+  const willRetryWrite =
+    topology.s.options.retryWrites &&
+    !inTransaction &&
+    supportsRetryableWrites(server) &&
+    operation.canRetryWrite;
 
-    if (
-      originalError instanceof MongoNetworkError &&
-      session?.isPinned &&
-      !session.inTransaction() &&
-      operation.hasAspect(Aspect.CURSOR_CREATING)
-    ) {
-      // If we have a cursor and the initial command fails with a network error,
-      // we can retry it on another connection. So we need to check it back in, clear the
-      // pool for the service id, and retry again.
-      session.unpin({ force: true, forceClear: true });
+  const hasReadAspect = operation.hasAspect(Aspect.READ_OPERATION);
+  const hasWriteAspect = operation.hasAspect(Aspect.WRITE_OPERATION);
+
+  if ((hasReadAspect && willRetryRead) || (hasWriteAspect && willRetryWrite)) {
+    if (hasWriteAspect && willRetryWrite) {
+      operation.options.willRetryWrite = true;
+      session.incrementTransactionNumber();
     }
+  }
 
-    // select a new server, and attempt to retry the operation
-    topology.selectServer(selector, serverSelectionOptions, (error?: Error, server?: Server) => {
-      if (!error && isWriteOperation && !supportsRetryableWrites(server)) {
-        return callback(
-          new MongoUnexpectedServerResponseError(
-            'Selected server does not support retryable writes'
-          )
-        );
+  try {
+    try {
+      return await operation.executeAsync(server, session);
+    } catch (operationError) {
+      if ((hasReadAspect && willRetryRead) || (hasWriteAspect && willRetryWrite)) {
+        if (operationError instanceof MongoError) {
+          return await retryOperation(operation, operationError, {
+            session,
+            topology,
+            selector
+          });
+        }
       }
+      throw operationError;
+    }
+  } finally {
+    if (session?.owner != null && session.owner === owner) {
+      await session.endSession().catch(() => null);
+    }
+  }
+}
```

Comment:
in the original execute_operation, we only execute the operation when either `(hasReadAspect && willRetryRead) || (hasWriteAspect && willRetryWrite))`, otherwise we fall through the the call to execute that doesn't retry.  We need to preserve this behavior and move this try-catch into the above if-statement.  That also means we may want to flip the logic back so that we have the catch-all call to execute without retry at the end of the method.

Reply:
from slack: there is gating in the catch block, no change in behavior